guix: Add guix-{attest,verify} scripts

dongcarl opened the pull request:
Adds replacements for `gsign` and `gverify`.

Personally I'm not a big fan of using the word "sign" as it's been used to refer to both codesigning and GPG signing.

dongcarl committed:
guix: Add guix-attest script

dongcarl committed:
guix: Add guix-verify script

dongcarl commented:
Perhaps we can test this by doing a trial run? I've PR'ed my attestations of 5da0e361b8b5a6d730a244e774c6810946619229 here: https://github.com/bitcoin-core/guix.sigs/pull/1

What I did:
- Run `guix-build`
- Run `env GUIX_SIGS_REPO=<path/to/guix.sigs> SIGNER=0x96AB007F1A7ED999=dongcarl ./contrib/guix/guix-attest`

Once you've run `./contrib/guix/guix-attest`, you can verify against my signatures using:
```
$ env GUIX_SIGS_REPO=<path/to/guix.sigs> ./contrib/guix/guix-verify
```

hebasto reviewed (commented):
Tested 5da0e361b8b5a6d730a244e774c6810946619229.

Result: https://github.com/bitcoin-core/guix.sigs/pull/2

> Once you've run `./contrib/guix/guix-attest`, you can verify against my signatures using:
> 
> ```
> $ env GUIX_SIGS_REPO=<path/to/guix.sigs> ./contrib/guix/guix-verify
> ```

I think it is possible when https://github.com/bitcoin-core/guix.sigs/pull/1 is merged, no?

fanquake commented:
> I think it is possible when bitcoin-core/guix.sigs#1 is merged, no?

The PR doesn't have to be merged, you should be able to just fetch and checkout the branch you want to verify against.

fanquake commented:
I have done a trial run through the attestation / verify steps as well. https://github.com/bitcoin-core/guix.sigs/pull/3.

Note that for anyone attesting/verifying on macOS, the invocations of `find` and `xargs` used in the attestation script are not currently compatible with BSD `find` or `xargs`. I just swapped them out for `gfind` and `gxargs` for now.

sipa commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,170 @@
+            find . -type f -printf '%P\0' | env LC_ALL=C sort -z | xargs -r0 sha256sum >> "$outsigdir"/SHA256SUMS
+        )
+        echo "${outname}: Signing SHA256SUMS to produce SHA256SUMS.asc"
+        gpg --detach-sign --local-user "$gpg_key_name" --output "$outsigdir"/SHA256SUMS.asc "$outsigdir"/SHA256SUMS
```
If you're producing .asc files I think it would be expected that they're ascii-armored (`--armor`); otherwise use the .sig extension (this doesn't really matter, but it's a convention).

RonSherfey reviewed (changes requested):
requested change

laanwj commented:
> Personally I'm not a big fan of using the word "sign" as it's been used to refer to both codesigning and GPG signing.

I tend to agree, "gitian signature" has always been somewhat ambiguous and confusing, attestation is a better word.

laanwj commented:
FWIW: I have a multi-step flow:

- On the build VM, `gitian-sign` with `-p true` to generate `.assert` files but bypass the actual signing
- Copy `.assert` files to signing machine
- On signing machine, generate detached signatures and check them into git

I'm not sure how far it is reasonable to accommodate my specific way of working (though there may be more people doing this, it's good to be careful with signing keys). But it would be nice to be able to skip the `gpg` invocation in `guix-attest`.

For checking gitian signatures, I have been using my own script for this as well, see:
bitcoin-core/bitcoin-maintainer-tools#71 bitcoin-core/bitcoin-maintainer-tools#90 . I think I can straightforwardly adapt it to Guix attestations when needed.

dongcarl committed:
guix-attest: Only use cross-platform flags for find+xargs

dongcarl committed:
guix-attest: Use ascii-armor signatures

dongcarl committed:
guix-attest: Allow skipping GPG signing with NO_SIGN

dongcarl commented:
Pushed 5da0e361b8b5a6d730a244e774c6810946619229 -> 0c238fd118742d6c059dc114b5cc0747adf7f6c5
- Added commits to address:
  - https://github.com/bitcoin/bitcoin/pull/21462#issuecomment-817068909
  - https://github.com/bitcoin/bitcoin/pull/21462#discussion_r611083278
  - https://github.com/bitcoin/bitcoin/pull/21462#issuecomment-817857040

laanwj: Please let me know if the `NO_SIGN` flag is enough for your workflow!

laanwj commented:
> laanwj: Please let me know if the NO_SIGN flag is enough for your workflow!

Yes seems perfect!

laanwj commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,188 @@
+# Required non-builtin commands should be invokable
+################
+
+check_tools cat env basename mkdir xargs find gpg
```
`gpg` check can be skipped in case of `NO_SIGN`?

laanwj commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,188 @@
+    signer_name="$gpg_key_name"
+fi
+
+if ! gpg --dry-run --list-secret-keys "${gpg_key_name}" >/dev/null 2>&1; then
```
This check can be skipped in case of NO_SIGN

laanwj commented:
I succeeded in generating attestations and signing them externally after removing the above checks and
```bash
GUIX_SIGS_REPO=/home/guest/guix.sigs SIGNER=laanwj NO_SIGN=1 contrib/guix/guix-attest
```

My only comment really is that having one attestation file per architecture means having to create a lot of signatures. I need to push the button on my external signer device for every one of them :smile:

dongcarl commented:
Pushed 0c238fd118742d6c059dc114b5cc0747adf7f6c5 -> 816c70c3cef2dde50436461f3fe43f8915840a26
- Addressed https://github.com/bitcoin/bitcoin/pull/21462#pullrequestreview-635756951
- Addressed https://github.com/bitcoin/bitcoin/pull/21462#pullrequestreview-635757547

sipa commented:
Building 816c70c3cef2dde50436461f3fe43f8915840a26

laanwj commented:
Tested ACK 816c70c3cef2dde50436461f3fe43f8915840a26
Results at bitcoin-core/guix.sigs#4

sipa commented:
Tested ACK 816c70c3cef2dde50436461f3fe43f8915840a26. I tried building, attesting, and verifying (as well as testing that verification fails with invalid signature).

As a follow-up, I think the output of guix-verify could be improved a bit (giving a global success/fail output, and perhaps grouping the output by signer).

hebasto reviewed (commented):
Tested 816c70c3cef2dde50436461f3fe43f8915840a26: https://github.com/bitcoin-core/guix.sigs/pull/6

Tested both `contrib/guix/guix-attest` and `contrib/guix/guix-verify`.

sipa commented:
Actually, I think it would be better if the signed data contained a stronger commitment to the source code it was built from than the truncated git commit id.

Attestations are really claims by the builder that certain input source code corresponds to the resulting build output. As the gpg-signed SHA256SUMS file doesn't actually commit the source code, someone could try to grind git commit ids to a collide with honest source code, get signers to accidentally guix attest the malicious source code, and then distribute the malicious build output along with a claim it was built from honest source code.

A possibility would be including the dist-archive source tgz in every SHA256SUMS file.

dongcarl commented:
You're right, attesting to truncated git commit ids by themselves is problematic and we should attest to the source code as well. I'm wondering what improvement there would be to attesting to the source tgz in every SHA256SUMS file vs. attesting to it in `dist-archive/$SIGNER/SHA256SUMS` like we do now?

sipa commented:
@dongcarl You can selectively combine (dist-archive/$SIGNER/SHA256SUMS from the honest source code, binary-platform/$SIGNER/SHA256SUMS from the malicious one).

dongcarl commented:
Let's say we do attest to the source tgz in every SHA256SUMS file. If I were a malicious signer, I could still manually construct the SHA256SUMS to attest to honest source code and malicious binaries, no?

sipa commented:
@dongcarl You can't modify the SHA256SUMS without invalidating its signatures. The signatures are per-SHA256SUM, so you can mix and match different SHA256SUMS files, but you can't mix and match lines within one file.

dongcarl commented:
> They can then take the dist-archive SHA256SUMS + honest sigs on it from the good source code, and the x86_64 SHA256SUMS file + its (honestly created) sigs on the malicious code.

Okay I see... And if the malicious person combines them, they can make it seem like the honest signer is producing malicious code... Am I understanding that right?

sipa commented:
Right, they can then make a distribution of the software along with SHA256SUMS{,.asc} with malicious binaries and honest source code, signed by honest signers.

dongcarl commented:
Ah that makes sense! Thanks for spotting it and being patient with me. I will make the changes :-)

sipa commented:
So, if you'd want to do this exactly, I think you'd want to have a build process that first creates the dist-archive, and then afterwards builds the binaries, using just the dist-archive output as input (rather than the source tree you were working on).

Emzy commented:
Tested ACK.
Works with Yubikey as signing device.

dongcarl committed:
guix: Minor quoting fix in libexec/build.sh

dongcarl committed:
guix: Construct $OUTDIR in ${DISTSRC}/output

While files are being output to $OUTDIR, it will be under
${DISTSRC}/output, and only when everything is done, will
${DISTSRC}/output be moved to the actual $OUTDIR.

This makes it so that a Ctrl-C in the middle of a build is less likely
to result in a partially-constructed $OUTDIR. In fact, if I understand
correctly, if $OUTDIR and $DISTSRC reside on the same filesystem, the
move (rename) is likely atomic.

Also, since the "working $OUTDIR" is under ${DISTSRC}/output, it will be
cleaned properly by the guix-clean script.

dongcarl committed:
guix: Attest to inputs in inputs.SHA256SUMS

At build/codesigning-time, hash build inputs and output the digest to
${OUTDIR}/inputs.SHA256SUMS, which gets included in the final SHA256SUMS
constructed by guix-attest.

Example final SHA256SUMS:
ee832d2a35b7701bff581dea05a536118b118e3ad0a587a2855b6ee8cd6fba20  inputs/bitcoin-78199266af7b.tar.gz
ca765e70a0c12866dd63c0be228b675278a26329e5f8f5b5c52fd09200fedf21  bitcoin-78199266af7b-powerpc64le-linux-gnu-debug.tar.gz
dae95327d7f2c324e2728c4b73627be6cb2c0d2f2e5bea940d1d5e6463939327  bitcoin-78199266af7b-powerpc64le-linux-gnu.tar.gz

dongcarl commented:
Pushed 816c70c3cef2dde50436461f3fe43f8915840a26 -> 7cd57f3ac308ba296842289542b15e902f105964
- The build output is now constructed in $DISTSRC/output, then moved (likely atomically) to $OUTDIR when everything is done
- Inputs are now attested to, thanks to sipa's suggestions

Please see commit messages for more explanations

dongcarl commented:
Wondering if the latest changes look alright, and what else needs fixing up here before it's ready for merge!

MarcoFalke commented on contrib/guix/guix-attest:
```diff
@@ -153,10 +153,17 @@ for outdir in "${OUTDIRS[@]}"; do
+            if [ -e inputs.SHA256SUMS ]; then
+                echo "${outname}: Including existent input SHA256SUMS"
+                cat inputs.SHA256SUMS >> "$outsigdir"/SHA256SUMS
+            fi
```
shouldn't this fail when the file is missing?

dongcarl replied:
Thinking about this since `dist-archive` does not have an `inputs.SHA256SUMS`... Right now, we attest to the bitcoin repo's git-archive in `distsrc-archive`, but I think perhaps it only makes sense to attest to it in the context of it being inputs to the host-specific outputs. When we add codesignatures, similarly it only makes sense to attest to the detached-sigs repo's git-archive in the context of it being inputs to the host-specific codesigned outputs. Will make a change!

maflcko reviewed (commented):
Concept ACK 7cd57f3ac308ba296842289542b15e902f105964 🍠

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK 7cd57f3ac308ba296842289542b15e902f105964 🍠
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiWMwv+MSUhxaO+wyNQ7iSyL1e3PrBcw5kLsXZDz8aWT27W8ZMJLoArJWD6K10p
o5QHYUJSg9HL/4QOQ+g+VCt6fdA2ktjpeKsrKfoivgMwfV5ZO2UowUHazuaxzi/k
oAR+xOLaWpsuKqvwSEsV9n0E53A7a5E/6jj5O2U9Nt2QFStRCW82q5Sm6cApy/mn
wTrvHQn4cqFqe+Io2WKnIvLFP8V/blggXRbGynm8+v1PBGyVTXoBP/CVbklc7JNj
Wl2hcY2Fzn0LfWgqRagtAlOTH/xxwbvPm/6PMfOqJg0u2Y4/e6KojTuFBYEQrXM4
vrjI3rRqjzg6fMa3miSZAwfknCIQJbDX1u7QIie4cqDolVM0HyGZizFxIr3m6hsc
VQ0Wu+eCrHzB3K2UdI5lo4Nn8O+quCICmue5d9Iix1obNdxCR4Q915EMWIejjkkR
45e2IOwJOT0ADQy3Drd2X6CQAHZTHJGtaoPecFR6DRTUE2uKHTYiKLEEb4QYJHBC
1B6vN1+j
=bEwp
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bc7b8ef6e6110dba1bb1148d42a8626cc8670590bd75fcef192668f241327364  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bc7b8ef6e6110dba1bb1148d42a8626cc8670590bd75fcef192668f241327364f010d6d4174afcb009b60d2b6622872d33ae08fff0101e22fed22deabf5c98258e820e02926108f020cf0002568681f2157f014efc943c6b4c669a131d68e45babee819d27c974b68108f0204628409b271878a3e95f9c420e5632890c95dd396ca1f2160ec761bf21b29e7b08f104608bd908f0082cbfee73077e44020083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104f6b2749f98ec92ef652557131953b4408f104608bd907f008da685278ed0863ab0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df01089dc20e29f5617e63195b1014759d54608f104608bd908f0087352515310a9807e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

dongcarl committed:
guix: Skip attesting to dist-archive

We already attest to the relevant dist-archive in inputs.SHA256SUMS,
which is recorded at build-time.

We use a SKIPATTEST.TAG file to indicate output directories which do not
require attestation (much like the CACHEDIR.TAG specification).
Generally, it's better to have build scripts declare properties of
directories instead of introducing name-based special cases in attest
scripts since build scripts have a more detailed context of what is
going on.

dongcarl commented:
Pushed 7cd57f3ac308ba296842289542b15e902f105964 -> feda2c8e3180cb983c35976d4440cea23a155b7f
- Addressed: https://github.com/bitcoin/bitcoin/pull/21462#discussion_r623755390, https://github.com/bitcoin/bitcoin/pull/21462#discussion_r623770745 (see last commit)

As always, guix.sigs are up at: https://github.com/dongcarl/guix.sigs/tree/2021-04-PR21462/feda2c8e3180

dongcarl committed:
guix-attest: Avoid incomplete sigdirs with ERR traps

Sometimes GPG connects to the wrong agent... or you don't have your
smartcard handy...

dongcarl commented:
Pushed feda2c8e31..d420e5c1c0:
- Added an ERR trap to remove incomplete sigdirs if signing fails

guix.sigs up at: https://github.com/dongcarl/guix.sigs/tree/2021-04-PR21462/d420e5c1c015

laanwj commented:
Tested ACK
I get the same output.
```
for x in *; do diff -du $x/dongcarl/SHA256SUMS $x/laanwj/SHA256SUMS; done
```
Would it be useful to have a concatenated SHA256SUMS (with all files) as well? I mean, depending on how we're going to do the distribution on the website and torrent. 

Or is the idea is to start using a subdirectory per platform there (each with its own SHA256SUMS) too? In this case, having `inputs/bitcoin-d420e5c1c015.tar.gz` in every one might not be practical, it implies having to duplicate the tarball . Maybe `../bitcoin-d420e5c1c015.tar.gz` then have the tarball in the top-level directory?

dongcarl commented:
> Or is the idea is to start using a subdirectory per platform there (each with its own SHA256SUMS) too? In this case, having `inputs/bitcoin-d420e5c1c015.tar.gz` in every one might not be practical, it implies having to duplicate the tarball .

Hmmm, are you talking about for the webserver? I think it can be done using hard/symlinks at the FS level, or just using virtualized paths at the HTTP level.

Wondering what particular user-flow you have in mind :-)

laanwj commented:
Code review and tested ACK d420e5c1c015f58d07aca4d6a805086488f74d03

> Wondering what particular user-flow you have in mind :-)

After discussion on IRC: let's leave this for another PR.

dongcarl commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,207 @@
+            echo "${outname}: Hashing build outputs to produce SHA256SUMS"
+            files="$(find -L . -type f ! -iname '*.SHA256SUMS')"
+            if [ -n "$files" ]; then
+                cut -c3- <<< "$files" | env LC_ALL=C sort | xargs sha256sum >> "$outsigdir"/SHA256SUMS
```
Hmmm, does `shasum -a 256` produce identical output to `sha256sum`?

Sjors replied:
If my uploaded attestation is correct, yes :-)
